### PR TITLE
feat: enable Kafla ACL authorization checks for Pull Queries

### PR DESCRIPTION
**Pull request by spena**
### Description 
**Notice:** This feature is part of implementing a KSQL authorization cache to enable Pull Queries #4158
**Notice:** The first 5 commits of this PR are from #4184 and #4186. Please skip those 5 commits to be reviewed in this PR.

This PR enables KSQL authorization checks in Kafka for pull queries now that an authorization cache is impemented in https://github.com/confluentinc/ksql/pull/4186.

The only change done in this PR is to remove the validation that skips pull query authorization validation.

### Testing done 
* Modified unit tests
* Verified manually with CLI and a KSQL server with RBAC and ACL configurations.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by vinothchandar** [comment]
@spena does this PR subsume  #4184 and #4186. I am wondering if I can just review this one , if you are planning to land this with the commits from the other PRs.

File: ksql-engine/src/main/java/io/confluent/ksql/security/KsqlSecurityContext.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Returns a {@code java.security.Principal} object containing the name of the current
+   * authenticated user. If the user has not been authenticated, the method returns
+   * {@code Optional.empty}.
+   *
+   * @return a {@code java.security.Principal} containing the name of the user making this request;
+   * {@code Optional.empty} if the user has not been authenticated
+   */
+  public Optional<Principal> getUserPrincipal() {
```
Comment (vinothchandar) [inline ksql-engine/src/main/java/io/confluent/ksql/security/KsqlSecurityContext.java]: reading the PR again.. just want to confirm that the main purpose of introducing `KsqlSecurityContext` is to have the user name part of the caching key?

Comment (spena) [reply ksql-engine/src/main/java/io/confluent/ksql/security/KsqlSecurityContext.java]: Yes.

I initially wanted to wrap the user name into the `ServiceContext`. But that class has a constructor with too many parameters, and I didn't want to add a new one. Also, seems having a `KsqlSecurityContext` might be beneficial in the future if we want to add extra security configurations there, such as ssl certificates, session tokens, etc.

**Comment by spena** [comment]
@vinothchandar This PR is ready for review.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java [hunk 2]
```diff
@@ -304,44 +302,24 @@ public void shouldNotCreateExternalClientsForPullQuery() {
...
-    assertThat(response.getEntity(), is(instanceOf(KsqlErrorMessage.class)));
-    final KsqlErrorMessage expectedEntity = (KsqlErrorMessage) response.getEntity();
-    assertThat(
-        expectedEntity.getMessage(),
-        containsString(KsqlConfig.KSQL_PULL_QUERIES_SKIP_ACCESS_VALIDATOR_CONFIG)
-    );
-  }
-
-  @Test
-  public void shouldPassCheckForPullQueryIfNotValidating() {
```
Comment (vinothchandar) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: is there a test to ensure the backend auth is not invoked if validation is not required? i.e non-rbac/acl case with insecure clusters

Comment (spena) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: Those tests are executed in other classes, like in the cache. I don't think we need that test case in this place.
I will merge this PR and add more tests if I find more.

**Comment by spena** [comment]
Jenkins is still failing with an unrelated error during packaging. All tests passed. I will merge this.

**Pull request merged**